core/src/main/scala/org/http4s/EntityDecoder.scala

```diff
@@ -106,11 +106,19 @@ object EntityDecoder extends EntityDecoderInstances {
       msg.headers.get(`Content-Type`) match {
         case Some(contentType) =>
           if (a.matchesMediaType(contentType.mediaType)) a.decode(msg, strict)
-          else b.decode(msg, strict)
+          else b.decode(msg, strict).leftMap{
+            case MediaTypeMismatch(actual, expected) =>
+              MediaTypeMismatch(actual, expected ++ a.consumes)
+            case other => other
+          }
 
         case None =>
           if (a.matchesMediaType(UndefinedMediaType)) a.decode(msg, strict)
-          else b.decode(msg, strict)
+          else b.decode(msg, strict).leftMap{
+            case MediaTypeMissing(expected) =>
+              MediaTypeMissing(expected ++ a.consumes)
```

Comment:
It bothers me ever so slightly that the type system cannot guarantee our expectation to only receive a `MediaTypeMismatch` when we have `Some(contentType)` and only receive a `MediaTypeMissing` when we have `None`(contentType). I wonder if this means that `UndefinedMediaType` is not the best thing and we would be better served with a `matchesUndefinedMediaType` method on `EntityDecoder`?